### PR TITLE
Document the technical design of incomplete type metadata

### PR DESCRIPTION
**Pull request by rjmccall**

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please smoke test and merge.

File: docs/ABI/TypeMetadata.rst [hunk 1]
```diff
@@ -85,14 +88,18 @@ contain the following fields:
 
 - The `nominal type descriptor`_ is referenced at **offset 1**.
 
-- A reference to the **parent** metadata record is stored at **offset 2**. For
-  enums that are members of an enclosing nominal type, this is a reference to
-  the enclosing type's metadata. For top-level enums, this is null.
+- If the enum is generic, then the
+  `generic argument vector`_ begins at **offset 2**.
+
+- In certain circumstances, there is a `payload size`_ immediately following
```
Comment (slavapestov) [inline docs/ABI/TypeMetadata.rst]: I think 'certain circumstances' are the enum has more than one payload case, and at least one payload case has a size that depends on a generic parameter or is resilient.

File: docs/ABI/TypeMetadata.rst [hunk 2]
```diff
@@ -477,3 +509,236 @@ contains:
...
+metadata.
+
+Specifically, a type metadata record is required to be complete when:
+
+- It is passed as a generic argument to a function (other than a metadata
+  access function, witness table access function, or metadata initialization
+  function).
+- It is stored as a metatype value.
+- It is used to build an existential value.
+- It is passed as the ``Self`` argument to a ``static`` or ``class`` method,
```
Comment (slavapestov) [inline docs/ABI/TypeMetadata.rst]: What about when it is passed as a non-self metatype argument?

File: docs/ABI/TypeMetadata.rst (hunk 1 above)
Comment (rjmccall) [reply docs/ABI/TypeMetadata.rst]: Do you think that's an important enough detail to include here?  I thought it was obscure enough to just leave it vague.  Honestly, I think it's arguably a private implementation detail more than something that needs documentation.

File: docs/ABI/TypeMetadata.rst (hunk 2 above)
Comment (rjmccall) [reply docs/ABI/TypeMetadata.rst]: "It is stored as a metatype value".  Maybe "stored" is too leading.

**Comment by rjmccall** [comment]
@slavapestov @jckarter I'm going on vacation next week, and I'd like to make sure that people understand how this works just in case something comes up before I disappear.  Can you let me know if the new section makes sense enough that you'd feel comfortable with the implementation?

**Pull request merged**